crates/bevy_ecs/src/query/access.rs

```diff
@@ -303,6 +303,28 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
         &mut self.access
     }
 
+    /// Returns a reference to the underlying filter sets, collected into disjunctive normal form.
+    ///
+    /// Filter sets are stored in an "ORs of ANDs" format, meaning that a filter `(With<A>, Or<(With<B>, Without<C>)>)`
+    /// expands into `Or<((With<A>, With<B>), (With<A>, Without<C>))>`. All [`AccessFilters`] in the filter sets vector
+    /// are implicitly joined with an `Or` clause, and each `AccessFilters` stores a set of `With` and `Without`
+    /// constraints.
+    #[inline]
+    pub fn filter_sets(&self) -> &Vec<AccessFilters<T>> {
+        &self.filter_sets
+    }
+
+    /// Returns a mutable reference to the underlying filter sets, collected into disjunctive normal form.
+    ///
+    /// Filter sets are stored in an "ORs of ANDs" format, meaning that a filter `(With<A>, Or<(With<B>, Without<C>)>)`
+    /// expands into `Or<((With<A>, With<B>), (With<A>, Without<C>))>`. All [`AccessFilters`] in the filter sets vector
+    /// are implicitly joined with an `Or` clause, and each `AccessFilters` stores a set of `With` and `Without`
+    /// constraints.
+    #[inline]
+    pub fn filter_sets_mut(&mut self) -> &mut Vec<AccessFilters<T>> {
```

Comment:
Can this method be exposed as a safe Rust function?

Reply:
I am concerned that this would allow an implementor of WorldQuery to hit UB in the case that:
1. The first term of a query is &T and so adds With<T> to the filters
2. The second term (since it now has mutable access to the filter sets in update_component_access) uses the new methods to remove the With<T> requirement.
3. You iterate the query and match an archetype without T hitting UB
Technically implementing WorldQuery is unsafe so allowing this wouldn't be breaking any guarantees, but I'm not sure what the use cases for direct access to the filters are that would make it worth exposing it.

Reply:
Yeah, that'd be UB. A `FilteredAccess:iter_filter_sets_mut` method would be safe and still fit my use case.

Reply:
This is safer but you can still edit the filters to remove requirements. What use case are you servicing here?

Reply:
Actually, in retrospect, even a mutable iterator could easily allow UB (i.e. by accidentally narrowing a filter with `AccessFilters::add_with` or loosening it with `add_without`, instead of using the safe `and_with` and `and_without` methods)...